### PR TITLE
Add hidden `!botlog` command to fetch cloudwatch logs

### PR DESCRIPTION
**Pull request by joshblum**

### CODE REVIEW

File: base/base.go [hunk 1]
```diff
@@ -29,20 +30,28 @@ type Server struct {
...
+	shutdownCh         chan struct{}
+	announcement       string
+	awsRegion          string
+	cloudwatchlogGroup string
+	kbc                *kbchat.API
+	botAdmins          []string
 }
 
-func NewServer(announcement string) *Server {
+func NewServer(announcement, awsRegion, cloudwatchlogGroup string) *Server {
```
Comment (mmaxim) [inline base/base.go]: Make this part of an optional struct called `AWSOptions`

File: base/options.go [hunk 2]
```diff
@@ -1,19 +1,52 @@
...
+	fs.StringVar(&o.DSN, "dsn", os.Getenv("BOT_DSN"), "Bot database DSN")
+	if o.AWSOpts == nil {
+		o.AWSOpts = &AWSOptions{}
+	}
+	fs.StringVar(&o.AWSOpts.AWSRegion, "aws-region", os.Getenv("BOT_DSN"), "AWS region for cloudwatch logs, optional")
+	fs.StringVar(&o.AWSOpts.CloudWatchLogGroup, "cloudwatch-log-group", os.Getenv("BOT_CLOUDWATCH_LOG_GROUP"), "Cloudwatch log group name, optional")
+	if err := fs.Parse(argv[1:]); err != nil {
+		return err
+	}
+	if len(o.DSN) == 0 {
```
Comment (mmaxim) [inline base/options.go]: Every bot needs this? Doesn't seem right.

File: webhookbot/main.go [hunk 3]
```diff
@@ -120,20 +120,14 @@ func main() {
 }
 
 func mainInner() int {
-	var opts Options
-	flag.StringVar(&opts.KeybaseLocation, "keybase", "keybase", "keybase command")
-	flag.StringVar(&opts.Home, "home", "", "Home directory")
-	flag.StringVar(&opts.Announcement, "announcement", os.Getenv("BOT_ANNOUNCEMENT"),
-		"Where to announce we are running")
-	flag.StringVar(&opts.HTTPPrefix, "http-prefix", os.Getenv("BOT_HTTP_PREFIX"),
+	opts := &Options{}
```
Comment (mmaxim) [inline webhookbot/main.go]: Not a big fan of this, should have a `base.NewOptions()`

Comment (joshblum) [reply webhookbot/main.go]: yeah just adds a bunch of boilerplate unfortunately

File: base/options.go (hunk 2 above)
Comment (joshblum) [reply base/options.go]: mmm true

File: webhookbot/main.go (hunk 3 above)
Comment (mmaxim) [reply webhookbot/main.go]: This case here just seems like it is incorrect anyway, shouldn't it be `opts := &base.Options{}`

Comment (mmaxim) [reply webhookbot/main.go]: Oh wait nvm

**Comment by joshblum** [comment]
@mmaxim this is ready

File: githubbot/main.go [hunk 4]
```diff
@@ -233,24 +239,23 @@ func main() {
...
 	if len(opts.DSN) == 0 {
-		fmt.Printf("must specify a poll database DSN\n")
+		fmt.Printf("must specify a database DSN\n")
 		return 3
 	}
-	bs := NewBotServer(opts)
+	bs := NewBotServer(*opts)
 	if err := bs.Go(); err != nil {
 		fmt.Printf("error running chat loop: %v\n", err)
+		return 3
```
Comment (Daniel-Wang) [inline githubbot/main.go]: @joshblum did you want to `return 3` after all the other `fmt.Printf("error running chat loop: %v\n", err)` too?
Seems like only in githubbot this change exists.
Was applying some retroactive changes to gitlabbot so I was looking at the PR XD

Comment (joshblum) [reply githubbot/main.go]: yes that sounds good

**Pull request merged**